File: LeetCode/longest_valid_parenthesis.rb
class String
    def lvp
        stack = []
        last = 0
        max = 0
        0.upto(self.size - 1).each { |i|
            if self[i] == '('
                stack.push(i)
            else
                if !stack.any?
                    last = i
                else
                    stack.pop
                    max = stack.any?? [max, i - stack[-1]].max : [max, i - last].max
                end
            end
        }

        max
    end
end

[")()())", ")()"].each { |x|
    puts "#{x} - #{x.lvp}"
}

File: LeetCode/longest_valid_parenthesis.cs
using System;
using System.Collections.Generic;

class Program {
    static int Max(char[] p) {
        Stack<int> stack = new Stack<int>();
        int last = 0;
        int max = 0;
        for (int i = 0; i < p.Length; i++) {
            if (p[i] == '(') {
                stack.Push(i);
            } else {
                if (stack.Count == 0) {
                    last = i;
                } else {
                    stack.Pop();
                    if (stack.Count == 0) {
                        max = Math.Max(max, i - last);
                    } else {
                        max = Math.Max(max, i - stack.Peek());
                    }
                }
            }
        }

        return max; 
    }

    static void Main() {
        char[][] parenthesis = new char[][]{ new char[] {')', '(', ')', '(', ')', ')'}, new char[] {'(', '(', ')'}};
        foreach (var p in parenthesis) {
            Console.WriteLine("{0} - {1}", String.Join(",", p), Max(p));
        }
    }
}
